### ISSUE TITLE
no-unused-expressions fails with logical expressions

### ISSUE THREAD

**Issue by ev0rtex** [eslint/eslint#2418]
NOTE: Here is the issue for my pull request (#2417) I submitted - I went back and edited my commit message with this issue number.

When run against code like the following, the no-unused-expressions rule fails even though this should be valid:
```js
fs.accessSync(cfgName) || (confFile = cfgName);
```

**Comment by nzakas** [comment]
Can you explain more about what you think the exceptions should be? Are you saying that if one side has an assignment, we shouldn't warn?

**Comment by ev0rtex** [comment]
That's not what I'm saying at all. In the referenced commit you can see that I just broke out the bit that determines whether a given expression is valid (the `isValidExpression` function) and then it applies the same rules to both sides of a logical operator, e.g.:

```js
if (node.type === "LogicalExpression") {
    return isValidExpression(node.left) && isValidExpression(node.right);
}
```

**Comment by michaelficarra** [comment]
This doesn't make any sense. If you can say that, then you can say the same thing about binary expressions:

```js
a() + b();
```

This rule is notifying you that the *result* of that expression is unused. Which it is. There should be no exception for this pattern.

**Comment by nzakas** [comment]
Yeah, @michaelficarra is right. The result of the expression is lost, which is what this rule is designed to warn about. If you're using these of patterns, you should probably disable the rule using an inline comment.

**Comment by jwarkentin** [comment]
Here's something to consider. It's useful in many cases to have warnings about the result of an expression being unused, but in some cases it's not good. It is common to see something like this in JavaScript libraries:

```js
function myfn(options) {
  options || (options = {some: 'default'});
}
```

It is also common to see things like:

```js
somevar || myfn();
```

Unfortunately, both of those result in a linting error. I don't think it makes sense to warn when there is an assignment or a function call involved in the expression as that is clearly not a noop. You cannot know that calling a function has no effect and you know that an assignment in an expression has a definite effect. It makes sense to warn when you have an expression that is obviously a noop such as `a + b`.

**Comment by jwarkentin** [comment]
One more thing. It doesn't make sense to throw a linting error with patterns like those I demonstrated when they are useful, harmless and very common. It also doesn't make sense to require disabling the rule as expressions that definitely result in a noop should in fact be warned about. There should be a way to have a useful rule against such expressions while still allowing such a common usage pattern.

If not through this rule then there should be a new rule created like this one but that allows assignments and functions in unassigned expressions. I suspect though that such a rule would be preferred by most over this rule. Maybe I'm wrong about that, but it's worth consideration and community discussion at the very least.

**Comment by ilyavolodin** [comment]
@jwarkentin This rule was specifically designed to catch the patterns that you are arguing for. If you look at the documentation for the rule here: http://eslint.org/docs/rules/no-unused-expressions you will see that 5th example is exactly the same as what you tried to fix. Not everyone has to enable every rule. If you use patterns that goes against the rule - that's fine, there's nothing wrong with that. Just disable the rule. (in my own .eslintrc I only have a few rules enabled, out of 160 rules that we currently have). If you only like part of the rule, then propose a flag that would disable the part you don't like.

**Comment by nzakas** [comment]
@jwarkentin you can always disable the rule inline using a comment. That's what I'd recommend if you want to keep this rule on.
